src/main/java/org/opensearch/neuralsearch/sparse/data/SparseVector.java

```diff
@@ -0,0 +1,197 @@
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Sparse vector implementation, which is a list of (token, freq) pairs
+ */
+@EqualsAndHashCode
+public class SparseVector implements Accountable {
+    // tokens will be stored in order
+    private final short[] tokens;
+    private final byte[] weights;
+
+    public SparseVector(BytesRef bytesRef) throws IOException {
+        this(readToMap(bytesRef));
+    }
+
+    public int getSize() {
+        return tokens == null ? 0 : tokens.length;
+    }
+
+    public SparseVector(Map<String, Float> pairs) {
+        this(
+            pairs.entrySet()
+                .stream()
+                .map(t -> new Item(convertStringToInteger(t.getKey()), ByteQuantizer.quantizeFloatToByte(t.getValue())))
+                .collect(Collectors.toList())
+        );
+    }
+
+    private static Integer convertStringToInteger(String value) {
+        return NumberUtils.createInteger(value);
+    }
+
+    public SparseVector(List<Item> items) {
+        items.sort(Comparator.comparingInt(Item::getToken));
+        int size = items.size();
+        this.tokens = new short[size];
+        this.weights = new byte[size];
+        for (int i = 0; i < size; ++i) {
+            this.tokens[i] = (short) items.get(i).getToken();
+            this.weights[i] = items.get(i).getWeight();
+        }
+    }
+
+    private static Map<String, Float> readToMap(BytesRef bytesRef) {
+        Map<String, Float> map = new HashMap<>();
+        try (
+            ByteArrayInputStream bais = new ByteArrayInputStream(
+                ArrayUtil.copyOfSubArray(bytesRef.bytes, bytesRef.offset, bytesRef.length)
+            );
+            DataInputStream dis = new DataInputStream(bais)
+        ) {
+            while (bais.available() > 0) {
+                int stringLength = dis.readInt();
+                byte[] stringBytes = new byte[stringLength];
+                dis.readFully(stringBytes);
+                String key = new String(stringBytes, StandardCharsets.UTF_8);
+                float value = dis.readFloat();
+                map.put(key, value);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        return map;
+    }
+
+    public byte[] toDenseVector() {
+        int size = getSize();
+        if (size == 0) {
+            return new byte[0];
+        }
+        int maxToken = this.tokens[size - 1];
+        byte[] denseVector = new byte[maxToken + 1];
+        for (int i = 0; i < size; ++i) {
+            denseVector[this.tokens[i]] = this.weights[i];
+        }
+        return denseVector;
+    }
+
+    public int dotProduct(final byte[] denseVector) {
+        int score = 0;
+        int size = getSize();
+
+        // Early exit for empty vectors
+        if (size == 0 || denseVector.length == 0) return 0;
```

Comment:
I see several places there's no null check, I understand these methods are low level methods and rarely receiving null parameters, but is it better to add null check or at least using `assert`?